lgc/patch/RegisterMetadataBuilder.cpp

```diff
@@ -0,0 +1,1380 @@
+
+    if (m_pipelineState->isTessOffChip())
+      getHwShaderNode(Util::Abi::HardwareStage::Gs)[Util::Abi::HardwareStageMetadataKey::OffchipLdsEn] = true;
+  } else {
+    if (vsBuiltInUsage.instanceIndex)
+      esVgprCompCnt = 3; // Enable instance ID
+  }
+  getGraphicsRegNode()[Util::Abi::GraphicsRegisterMetadataKey::EsVgprCompCnt] = esVgprCompCnt;
+
+  // VGT_GS_MAX_VERT_OUT
+  const auto &geometryMode = m_pipelineState->getShaderModes()->getGeometryShaderMode();
+  unsigned maxVertOut = std::max(1u, static_cast<unsigned>(geometryMode.outputVertices));
+  getGraphicsRegNode()[Util::Abi::GraphicsRegisterMetadataKey::VgtGsMaxVertOut] = maxVertOut;
+
+  // VGT_GS_MODE
+  auto vgtGsMode = getGraphicsRegNode()[Util::Abi::GraphicsRegisterMetadataKey::VgtGsMode].getMap(true);
+  vgtGsMode[Util::Abi::VgtGsModeMetadataKey::Mode] = GS_SCENARIO_G;
+  vgtGsMode[Util::Abi::VgtGsModeMetadataKey::EsWriteOptimize] = false;
+  if (m_pipelineState->isGsOnChip()) {
+    vgtGsMode[Util::Abi::VgtGsModeMetadataKey::Onchip] = VGT_GS_MODE_ONCHIP_ON;
+    vgtGsMode[Util::Abi::VgtGsModeMetadataKey::GsWriteOptimize] = false;
+  } else {
+    vgtGsMode[Util::Abi::VgtGsModeMetadataKey::Onchip] = VGT_GS_MODE_ONCHIP_OFF;
+    vgtGsMode[Util::Abi::VgtGsModeMetadataKey::GsWriteOptimize] = true;
+  }
+  unsigned cutModeVal = GS_CUT_1024__HASHWVS;
+  if (geometryMode.outputVertices <= 128)
+    cutModeVal = GS_CUT_128__HASHWVS;
+  else if (geometryMode.outputVertices <= 256)
+    cutModeVal = GS_CUT_256__HASHWVS;
+  else if (geometryMode.outputVertices <= 512)
+    cutModeVal = GS_CUT_512__HASHWVS;
+  vgtGsMode[Util::Abi::VgtGsModeMetadataKey::CutMode] = cutModeVal;
+
+  // VGT_GS_ONCHIP_CNTL
+  auto vgtGsOnChipCntl = getGraphicsRegNode()[Util::Abi::GraphicsRegisterMetadataKey::VgtGsOnchipCntl].getMap(true);
+  vgtGsOnChipCntl[Util::Abi::VgtGsOnchipCntlMetadataKey::EsVertsPerSubgroup] = calcFactor.esVertsPerSubgroup;
+  vgtGsOnChipCntl[Util::Abi::VgtGsOnchipCntlMetadataKey::GsPrimsPerSubgroup] = calcFactor.gsPrimsPerSubgroup;
+  // NOTE: The value of field "GS_INST_PRIMS_IN_SUBGRP" should be strictly equal to the product of
+  // VGT_GS_ONCHIP_CNTL.GS_PRIMS_PER_SUBGRP * VGT_GS_INSTANCE_CNT.CNT.
+  const unsigned gsInstPrimsInSubgrp =
+      geometryMode.invocations > 1 ? (calcFactor.gsPrimsPerSubgroup * geometryMode.invocations) : 0;
+  vgtGsOnChipCntl[Util::Abi::VgtGsOnchipCntlMetadataKey::GsInstPrimsPerSubgrp] = gsInstPrimsInSubgrp;
+
+  // VGT_GS_VERT_ITEMSIZE and VGT_GSVS_RING_OFFSET
+  const unsigned itemCount = 4;
+  unsigned gsVertItemSizes[itemCount] = {};
+  unsigned gsVsRingOffsets[itemCount] = {};
+  unsigned gsVsRingOffset = 0;
+  for (unsigned i = 0; i < itemCount; ++i) {
+    gsVertItemSizes[i] = sizeof(unsigned) * gsInOutUsage.gs.outLocCount[i];
+    gsVsRingOffset += gsVertItemSizes[i] * maxVertOut;
+    gsVsRingOffsets[i] = gsVsRingOffset;
+  }
+  getGraphicsRegNode()[Util::Abi::GraphicsRegisterMetadataKey::VgtGsVertItemsize] = gsVertItemSizes;
```

Comment:
This and the next line give a warning "address of array will always evaluate to true" for me. Do you want to assign the whole array to a msgpack array node? There is no operator= for that; you'll have to do it manually.

Reply:
Yes, I made a wrong use for the array node. They are fixed in [d061491](https://github.com/GPUOpen-Drivers/llpc/commit/d0614913041bd72c1f3999d373fb19323ee72b2d)